### PR TITLE
add oauth interface

### PR DESCRIPTION
**Pull request by nsklikas**
The `oauth` interface will be provided by any charm that claims to implement an OAuth2/OIDC provider or required by any charm that wishes to register an OAuth2/OIDC client with the provider.

The library is not yet published, but, when it is, it will be at that link.

### CODE REVIEW

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:1]: Why do we need the `_client` suffix?

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:3]: Please add an Overview heading and provide a brief (1 paragraph max, 1 sentence ideally) overview of the relation. This will be required for all interfaces going forward.

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:1]: Why do we need the `_client` suffix? My general standpoint is that interface names should be kept concise and descriptive. Given that the provider of this relation is not even the client, but the oidc provider.

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:22]:
```suggestion
The requirer and the provider must adhere to a certain set of criteria to be considered compatible with the interface.
```

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:39]: Change format to "Is expected to..."

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:43]:
```suggestion
```

File: interfaces/oauth_client/v0/schemas/provider.json
Comment (simskij) [inline interfaces/oauth_client/v0/schemas/provider.json:1]: Rework these schemas to be at the root level, ie. contain a property for the application databag, and one property for the unit databag. See https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/prometheus_scrape/v0/schemas/provider.json for an example. This will eventually need to be done for all schemas.

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:30]: This means it wont be compatible with Juju 2.9. Is that an OK tradeoff? I think if so, this should be mentioned somewhere.

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [inline interfaces/oauth_client/v0/README.md:53]:
```suggestion
```

**Review by simskij — CHANGES_REQUESTED** [review]

File: interfaces/oauth_client/v0/README.md
Comment (nsklikas) [reply interfaces/oauth_client/v0/README.md]: Is the `Usage` section still required? It feels like the `Overview` section was included in the `Usage` section so far.

File: interfaces/oauth_client/v0/README.md
Comment (gruyaume) [reply interfaces/oauth_client/v0/README.md]: Since we are starting from scratch with IAM, I don't know why we would want to support juju 2.x.

Comment (gruyaume) [reply interfaces/oauth_client/v0/README.md]: I agree it should be mentioned somewhere to make it super clear

**Review by simskij — APPROVED** [review]

File: interfaces/oauth_client/v0/README.md
Comment (simskij) [reply interfaces/oauth_client/v0/README.md]: Yes, I think they serve different purposes. Usage section is developer documentation on how to use the interface. Think of overview as the "executive summary".

**Comment by jnsgruk** [comment]
This is probably ready to merge, but only once a link to the interface has been added to the top level README table, probably under a new "Identity" subsection

**Comment by simskij** [comment]
In general, I'd prefer if we could keep some good PR hygiene and resolve conversations that have been resolved, so it's easy to see when everything's been acted out.

**Review by jnsgruk — APPROVED** [review]
Thanks for the iteration @nsklikas :)

**Comment by nsklikas** [comment]
> In general, I'd prefer if we could keep some good PR hygiene and resolve conversations that have been resolved, so it's easy to see when everything's been acted out or addressed. 👍🏼

I assume that means that I should resolve those conversations, so I went ahead and resolved them. I believed it should be the commenter's responsibility to resolve their comments, so that we can be sure that everything has been addressed properly. That's why I left them open. 

It should be fine now, but I would rather keep this PR open until the actual code/library has been merged/published just to be sure that nothing is missing from this interface.

**Pull request merged**
